src/SubscribedStreamPartition.js

```diff
@@ -1,92 +1,102 @@
     constructor(client, streamId, streamPartition) {
         this._client = client
         this.streamId = streamId
         this.streamPartition = streamPartition
         this.subscriptions = {}
-        this.isPublisherPromises = {}
+        this.getStream = memoize(this.getStream.bind(this), memoizeOpts)
+        this.validator = new StreamMessageValidator({
+            getStream: this.getStream,
+            isPublisher: memoize(async (publisherId, _streamId) => (
+                this._client.isStreamPublisher(_streamId, publisherId)
+            ), memoizeOpts),
+            isSubscriber: memoize(async (ethAddress, _streamId) => (
+                this._client.isStreamSubscriber(_streamId, ethAddress)
+            ), memoizeOpts),
+        })
+        this.getPublishers = memoize(this.getPublishers.bind(this), memoizeOpts)
+        this.getSubscribers = memoize(this.getSubscribers.bind(this), memoizeOpts)
+        this.isValidPublisher = memoize(this.isValidPublisher.bind(this), memoizeOpts)
+        this.isValidSubscriber = memoize(this.isValidSubscriber.bind(this), memoizeOpts)
     }
 
     async getPublishers() {
-        if (!this.publishersPromise || (Date.now() - this.lastUpdated) > PUBLISHERS_EXPIRATION_TIME) {
-            this.publishersPromise = this._client.getStreamPublishers(this.streamId).then((publishers) => {
-                const map = {}
-                publishers.forEach((p) => {
-                    map[p] = true
-                })
-                return map
+        const publishers = await this._client.getStreamPublishers(this.streamId)
+        return publishers.reduce((obj, key) => (
+            Object.assign(obj, {
+                [key]: true
             })
-            this.lastUpdated = Date.now()
-        }
-        return this.publishersPromise
+        ), {})
     }
 
-    async _isPublisher(publisherId) {
-        if (!this.isPublisherPromises[publisherId]) {
-            this.isPublisherPromises[publisherId] = this._client.isStreamPublisher(this.streamId, publisherId)
-        }
-        return this.isPublisherPromises[publisherId]
+    async getSubscribers() {
+        const subscribers = await this._client.getStreamSubscribers(this.streamId)
+        return subscribers.reduce((obj, key) => (
+            Object.assign(obj, {
+                [key]: true
+            })
+        ), {})
     }
 
     async isValidPublisher(publisherId) {
-        const cache = await this.getPublishers()
-        if (cache[publisherId]) {
-            return cache[publisherId]
+        return this._client.isStreamPublisher(this.streamId, publisherId)
+    }
+
+    async isValidSubscriber(ethAddress) {
+        return this._client.isStreamSubscriber(this.streamId, ethAddress)
+    }
+
+    async _verifyStreamMessage(msg) {
+        try {
+            await this.validator.validate(msg)
+        } catch (err) {
+            // store error for possible introspection later
```

Comment:
What if this just threw on validation failure, instead of returning a boolean?

Reply:
This actually worked quite well, since it allows the actual error message to bubble up, rather than be wrapped. Note the old code differentiated by `Error` type between signature errors and validation errors, but new code does not, only changes error message.